routers/web/auth/oauth.go

```diff
@@ -685,7 +685,7 @@ func handleAuthorizationCode(ctx *context.Context, form forms.AccessTokenForm, s
 		})
 		return
 	}
-	if !app.ValidateClientSecret([]byte(form.ClientSecret)) {
+	if app.Confidential && !app.ValidateClientSecret([]byte(form.ClientSecret)) {
```

Comment:
Why this ValidateClientSecret is skipped?

Reply:
> NOT RECOMMENDED for authorization servers to require client authentication of public native apps clients using a shared secret

https://datatracker.ietf.org/doc/html/rfc8252#section-8.5

I'll add a comment.

Reply:
Hmm, it's just not recommended. According to 2.3

```
   The authorization server MAY establish a client authentication method
   with public clients.  However, the authorization server MUST NOT rely
   on public client authentication for the purpose of identifying the
   client.
```

It's still feasible to require the client secret for a public client. What's the reason that this behavior is a blocker?

(I am not an export for OAuth eco-system, I am still learning the details)

Reply:
True. Nevertheless GitHub and GitLab don't require the secret so I'd rather clients didn't have to special case Gitea.

Instead public clients are required to use PKCE  https://datatracker.ietf.org/doc/html/rfc8252#section-8.1 . This was missing from my first commit. 

Added missing logic.

Reply:
@wxiaoguang Here's a simpler fix for the client secret check problem https://github.com/go-gitea/gitea/pull/21351

Reply:
Reverted change